src/commands/containers/browseContainer.ts

```diff
@@ -0,0 +1,114 @@
+
+async function captureBrowseCancelStep<T>(cancelStep: ConfigureBrowseCancelStep, properties: BrowseTelemetryProperties, prompt: () => Promise<T>): Promise<T> {
+    return await captureCancelStep(cancelStep, properties, prompt)
+}
+
+// NOTE: These ports are ordered in order of preference.
+const commonWebPorts = [
+    443,    // SSL
+    80,     // HTTP
+    3000,   // (Node.js) Express.js
+    3001,   // (Node.js) Sails.js
+    5001,   // (.NET Core) ASP.NET SSL
+    5000,   // (.NET Core) ASP.NET HTTP
+    8080,   // (Node.js)
+    8081    // (Node.js)
+];
+
+const commonSslPorts = [
+    443,    // SSL
+    5001    // (.NET Core) ASP.NET SSL
+];
+
+function parsePortAndProtocol(portAndProtocol: string): { port: number, protocol: string | undefined } | undefined {
+    const splitPortAndProtocol = portAndProtocol.split('/');
+
+    if (splitPortAndProtocol.length > 0) {
+        const port = Number.parseInt(splitPortAndProtocol[0], 10);
+
+        if (!Number.isNaN(port)) {
+            return {
+                port,
+                protocol: splitPortAndProtocol.length > 1 ? splitPortAndProtocol[1].toLowerCase() : undefined
+            };
+        }
+    }
+
+    return undefined;
+}
+
+export async function browseContainer(context: IActionContext, node?: ContainerTreeItem): Promise<void> {
+    const telemetryProperties = <BrowseTelemetryProperties>context.telemetry.properties;
+
+    if (!node) {
+        node = await captureBrowseCancelStep('node', telemetryProperties, () => ext.containersTree.showTreeItemPicker<ContainerTreeItem>(ContainerTreeItem.allContextRegExp, context));
+    }
+
+    const inspectInfo = await node.getContainer().inspect();
+
+    const ports = inspectInfo.NetworkSettings && inspectInfo.NetworkSettings.Ports || {};
+    const possiblePorts =
+        Object.keys(ports)
+              .map(portAndProtocol => ({ mappings: ports[portAndProtocol], containerPort: parsePortAndProtocol(portAndProtocol) }))
+              // Exclude port designations we cannot recognized...
+              .filter(port => port.containerPort !== undefined)
+              // Exclude ports that are non-TCP-based (which should be the default)...
+              .filter(port => port.containerPort.protocol === undefined || port.containerPort.protocol === 'tcp')
+              // Exclude ports not mapped to the host...
+              .filter(port => port.mappings && port.mappings.length > 0);
+
+    telemetryProperties.possiblePorts = possiblePorts.map(port => port.containerPort.port);
+
+    if (possiblePorts.length === 0) {
+        // tslint:disable-next-line: no-floating-promises
+        ext.ui.showWarningMessage('No valid ports were mapped from the container to the host.');
+
+        return;
+    }
+
+    // Prefer common port (in order of preference)...
+    let selectedPort = possiblePorts.find(port => commonWebPorts.find(commonPort => commonPort === port.containerPort.port) !== undefined);
```

Comment:
Sets may not respect ordering so they may not work, but if they do this might be an easier way to do the intersection: https://2ality.com/2015/01/es6-set-operations.html#intersection

Reply:
We should try to make better use of the `Set` and `Map` types, but I'm not sure I want to do that here (quite yet).  (The spec indicates that sets keep values in insertion order, but I've not yet played with them.) . In any case, I discovered that my logic was inverted here anyway, so at least fixed that.